modules/core/form/js/form_protection.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * @file
+ * Show warning when a user is about to navigate away from an unsaved, dirty form.
+ */
+(function ($, Drupal) {
+
+    "use strict";
+
+    Drupal.behaviors.form_protection = {
+
+        initialFormState: {},
+
+        attach: function (context) {
+            const behavior = Drupal.behaviors.form_protection;
+            let allowClick = false;
+            let formIsDirty = false;
+
+            // Get initial form state on initial load.
+            $("form.entity-or-quick-form-protected :input", context).each(function () {
+                behavior.initialFormState[$(this).attr("id")] = $(this).serialize();
+            });
+
+            // Let all form submit buttons through.
+            $("input[type='submit'], button[type='submit']").each(function() {
+                $(this).addClass("allow-submit");
+                $(this).click(function() {
+                    allowClick = true;
+                });
+            });
+
+            // Catch all links and buttons EXCEPT for "#" links.
+            $("a, button, input[type='submit']:not(.allow-submit), button[type='submit']:not(.allow-submit)")
+                .each(function() {
+                    $(this).click(function() {
+                        formIsDirty = behavior.checkForUnsavedChanges();
+                        // If the user clicked on a # link or the form is not dirty, let click through; otherwise,
+                        // return something other than undefined to trigger onbeforeunload.
+                        if (formIsDirty && $(this).attr("href") !== "#") {
+                            return 0;
+                        }
+                    });
+                });
+
+            // Handle backbutton, exit etc.
+            window.onbeforeunload = function () {
+                if (formIsDirty && !allowClick) {
+                    allowClick = false;
+                    return (Drupal.t("You have unsaved changes."));

```

Comment:
Probably worth noting that modern browsers don't seem to allow custom messages so this is usually equivalent to returning `true` (though I haven't tested doing so). https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#browser_compatibility